File: app/src/main/java/com/mutant/godutch/model/Friend.java
package com.mutant.godutch.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.IntDef;

/**
 * Created by Mutant on 2017/5/26.
 */

public class Friend implements Parcelable {

    String uid;
    String name;
    String proPicUrl;
    int needToPay;

    public static final int STATE_ACCEPTED = 0;
    public static final int STATE_NOT_BE_ACCEPTED = 1;
    public static final int STATE_BE_INVITED = 2;
    int state;

    public Friend() {
    }

    public Friend(String uid, String name, String proPicUrl) {
        this.uid = uid;
        this.name = name;
        this.proPicUrl = proPicUrl;
    }

    public String getName() {
        return name != null ? name : "";
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getProPicUrl() {
        return proPicUrl;
    }

    public void setProPicUrl(String proPicUrl) {
        this.proPicUrl = proPicUrl;
    }

    public int getNeedToPay() {
        return needToPay;
    }

    public void setNeedToPay(int needToPay) {
        this.needToPay = needToPay;
    }

    public String getUid() {
        return uid;
    }

    public void setUid(String uid) {
        this.uid = uid;
    }

    @IntDef({STATE_ACCEPTED, STATE_NOT_BE_ACCEPTED, STATE_BE_INVITED})
    public @interface State {
    }

    public void setState(@State int state) {
        this.state = state;
    }

    @State
    public int getState() {
        return state;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.uid);
        dest.writeString(this.name);
        dest.writeString(this.proPicUrl);
        dest.writeInt(this.needToPay);
        dest.writeInt(this.state);
    }

    protected Friend(Parcel in) {
        this.uid = in.readString();
        this.name = in.readString();
        this.proPicUrl = in.readString();
        this.needToPay = in.readInt();
        this.state = in.readInt();
    }

    public static final Creator<Friend> CREATOR = new Creator<Friend>() {
        public Friend createFromParcel(Parcel source) {
            return new Friend(source);
        }

        public Friend[] newArray(int size) {
            return new Friend[size];
        }
    };
}

File: app/src/main/java/com/mutant/godutch/model/Friend.kt
package com.mutant.godutch.model

import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.IntDef

/**
 * Created by Mutant on 2017/5/26.
 */

class Friend : Parcelable {

    var uid: String = ""
    var name: String = ""
    var photoUrl: String = ""
    var debt: Double = 0.0
    var state: Int = STATE_ACCEPTED
        @State set

    constructor()

    constructor(uid: String, name: String, photoUrl: String) {
        this.uid = uid
        this.name = name
        this.photoUrl = photoUrl
    }

    constructor(uid: String, name: String, photoUrl: String, debt: Double, @State state: Int) {
        this.uid = uid
        this.name = name
        this.photoUrl = photoUrl
        this.debt = debt
        this.state = state
    }

    constructor(friend: Friend) : this(friend.uid, friend.name, friend.photoUrl, friend.debt, friend.state)

    @IntDef(STATE_ACCEPTED.toLong(), STATE_NOT_BE_ACCEPTED.toLong(), STATE_BE_INVITED.toLong())
    annotation class State

    companion object {
        const val STATE_ACCEPTED = 0
        const val STATE_NOT_BE_ACCEPTED = 1
        const val STATE_BE_INVITED = 2

        @JvmField
        val CREATOR: Parcelable.Creator<Friend> = object : Parcelable.Creator<Friend> {
            override fun createFromParcel(source: Parcel): Friend = Friend(source)
            override fun newArray(size: Int): Array<Friend?> = arrayOfNulls(size)
        }
    }

    constructor(source: Parcel) : this(
            source.readString(),
            source.readString(),
            source.readString(),
            source.readDouble(),
            source.readInt()
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(uid)
        dest.writeString(name)
        dest.writeString(photoUrl)
        dest.writeDouble(debt)
        dest.writeInt(state)
    }
}
